### PR TITLE
Create "suggested tests" tool in `rustbuild`

### PR DESCRIPTION
**Pull request by Ezrashaw**
Not the claimed person in #97339 but:
I've done a very rough implementation of this feature in-tree. I'm very new to `rustc` development (outside of docs) so some help would be greatly appreciated. The UI of this new subcommand obviously will change and I need some mentoring with the `--run` flag.

r? @jyn514

### CODE REVIEW

File: src/bootstrap/flags.rs
Comment (jyn514) [inline src/bootstrap/flags.rs:181]: it looks like this got autoformatted or something like that; do you mind reverting everything except the new `suggest` line?

**Review by jyn514 — COMMENTED** [review]
Thanks for starting on this! It looks like this is just the scaffolding so far - adding a new subcommand seems reasonable, but I would ideally like this to be a separate tool, not part of bootstrap, so that the compile times are better.

Does that make sense, do you know how to get started with the suggestions part of it?

**Comment by Ezrashaw** [comment]
Ok, just a few things for clarification:
- What do you mean by as a subcommand for `x.py` but also an external tool? I think that the tool existing inside rustbuild allows it to better be able to control the build process (and avoid all the complexities of invoking shell commands).
- Obviously you probably have a better understanding of this than me, but why might my 50-odd line `suggest.rs` file have an adverse impact on compile times compared to the massive other rustbuild files?
- Assuming that my current code is not completely different from what you had in mind, what should I do now? How could I implement the `--run` flag?

Thanks for your help.

**Comment by jyn514** [comment]
> What do you mean by as a subcommand for `x.py` but also an external tool? I think that the tool existing inside rustbuild allows it to better be able to control the build process (and avoid all the complexities of invoking shell commands).

See how e.g. `tidy` is invoked. We have lots of existing libraries in bootstrap for running commands, actually building the separate tool won't be too much work in comparison.

> Obviously you probably have a better understanding of this than me, but why might my 50-odd line `suggest.rs` file have an adverse impact on compile times compared to the massive other rustbuild files?

It means you have to rebuild rustbuild every time you change the suggestions, i.e. you have to recompile those massive other files. If you put it in a separate tool, you only have to recompile that tool itself, which makes it faster to iterate.

I'm not sure what you mean by a `--run` flag. Do you mean testing the suggested tests first, before other tests? That's a stretch goal, I think we should land this PR first before doing that.

**Comment by Ezrashaw** [comment]
Ok, moving the tool now makes a lot of sense, do you think I should put it in `src/tools` or an external repo? Also do you have any suggestions for my actual suggestions code, maybe it should load suggestions from a file?

Thanks again for your patience.

**Comment by jyn514** [comment]
src/tools seems fine :) I don't have strong opinions on how the tool itself should be designed - either a configuration file or hard-coding the patterns seems fine since it's going to be quick to compile.

**Comment by Ezrashaw** [comment]
Ok, could you also give some feedback on the `suggest.rs` file which contains the actual stuff. Also if I load suggestions from a file, would you be happy with the tool being directly in `src/bootstrap` for simplicity?

File: src/bootstrap/suggest.rs
Comment (jyn514) [inline src/bootstrap/suggest.rs:13]: Please use `get_rust_lang_rust_remote` here instead of `origin`.

Hmm, running merge-base happens enough there should probably be a helper function for it; but this is fine for now.

File: src/bootstrap/suggest.rs
Comment (jyn514) [inline src/bootstrap/suggest.rs:25]: this also happens enough it could probably use a helper function

File: src/bootstrap/suggest.rs
Comment (jyn514) [inline src/bootstrap/suggest.rs:39]: this is fine for debugging, but please remove it before unmarking the PR as a draft. you can hide it behind `builder.verbose` if you like.

File: src/bootstrap/suggest.rs
Comment (jyn514) [inline src/bootstrap/suggest.rs:87]: Hmm, this reminds me a lot of https://github.com/rust-lang/rust/issues/97338. I wonder if we should use that approach here, have steps tell you have to rerun them. it's a *lot* of work unfortunately :/ but it seems unfortunate to duplicate it.

on the other hand that would couple this pretty tightly to bootstrap which I don't want to do; so ignore me, sorry.

File: src/bootstrap/suggest.rs
Comment (jyn514) [inline src/bootstrap/suggest.rs:139]:
```suggestion
    "src/librustdoc/*" => [
```

File: src/bootstrap/suggest.rs
Comment (jyn514) [inline src/bootstrap/suggest.rs:140]:
```suggestion
        sug!("test", 1, ["tidy", "rustdoc", "rustdoc-ui"])
```

File: src/bootstrap/suggest.rs
Comment (jyn514) [inline src/bootstrap/suggest.rs:129]:
```suggestion
    "library/std/*" => [
```

File: src/bootstrap/suggest.rs
Comment (jyn514) [inline src/bootstrap/suggest.rs:63]: 1. Add a new tool in `src/tools`; take a look at expand-yaml-anchors or something for ideas.
2. Add a new `Step` which knows how to build that tool and returns the path as its `Output`.
3. Use `builder.run(my_tool_path)`

**Review by jyn514 — COMMENTED** [review]
No, please put it in `src/tools`, src/bootstrap is specifically for rustbuild.

File: src/bootstrap/suggest.rs
Comment (Ezrashaw) [reply src/bootstrap/suggest.rs:63]: I'm not particularly keen on moving it into its own tool now. If the tool itself stays quite small and static but suggestions are dynamically loaded then it kind of seems unnecessarily complex but I'll move it into `src/tools` if you want.

This bit of code is really about actually running the suggestions, but that is addressed above? It seems pretty important to the whole tool for it to be able to run the suggested tests somehow (for CI or  whatever).

File: src/bootstrap/suggest.rs
Comment (Ezrashaw) [reply src/bootstrap/suggest.rs:139]: Obviously these suggestions are really just examples and have to be built on over time.

File: src/bootstrap/suggest.rs
Comment (Ezrashaw) [reply src/bootstrap/suggest.rs:13]: I can't find the function you are talking about? A lot of `src/bootstrap` seems to use hardcoded remotes.

**Comment by Ezrashaw** [comment]
@jyn514 I've applied most of your suggestions and I've written some (hacky) code for the `--run` flag. As far as I can tell, it is integrating seamlessly with the rest of the build process. Do you think I'm on the right track with this? Or is it not very good? I'll also implement dynamically loaded suggestions soon.

File: src/bootstrap/suggest.rs
Comment (jyn514) [reply src/bootstrap/suggest.rs:17]: It was added in https://github.com/rust-lang/rust/pull/105702, you probably need to rebase over master. https://github.com/rust-lang/rust/pull/106286 is moving it so you might want to wait until that's merged, `origin` is fine in the meantime.

File: src/bootstrap/builder.rs
Comment (jyn514) [inline src/bootstrap/builder.rs:870]: I think you can avoid needing this new function by modifying `build.config.cmd` and using `Builder::new` instead. You could change `sug!` to generate the command for you so it's not too much of a pain:
```rust
(test, $stage:literal, $paths:expr) => Suggestion { stage: $stage, cmd: Subcommand::Test { paths: $paths } },
```

**Review by jyn514 — COMMENTED** [review]
> Sorry I should have been more clear: I'm not particularly keen on moving it into its own tool now. If the tool itself stays quite small and static but suggestions are dynamically loaded then it kind of seems unnecessarily complex but I'll move it into src/tools if you want.

ah interesting! ok, that seems reasonable now that you're using reflection in bootstrap itself :) looks like you haven't started working on the dynamic suggestions yet - I was imaging that as a JSON file, something like this:
```
$ cat src/bootstrap/suggestions.json
{
  "path": "compiler/rustc_data_structures/*",
  "suggestions": [
    {"cmd": "check"},
    {"cmd": "test", "stage": 0, "paths": ["tidy", "compiler/rustc_data_structures"] }
  ]
}
```

File: src/bootstrap/builder.rs
Comment (Ezrashaw) [reply src/bootstrap/builder.rs:870]: Thanks for the idea, I think I also need to modify `build.config.stage`. Sorry I'm on holiday right now so the progress is slow.

**Comment by Ezrashaw** [comment]
Seems quite complex :P. I was thinking more like plain text if that's ok, because then it becomes easier to write tidy checks, etc. Assuming JSON, is there a library preference?

**Comment by jyn514** [comment]
Bootstrap already uses `serde_json`, you can reuse that :)

What would a text format look like? The fields you're parsing are pretty nested, I worry that using a non-structured format would be pretty ambiguous.

File: src/bootstrap/builder.rs
Comment (jyn514) [reply src/bootstrap/builder.rs:870]: No need to apologize! You're responding quite fast actually :smile:

**Comment by Ezrashaw** [comment]
Yes, that's a good point. Also it's probably easier to `#[derive(Deserialize)]` a bunch than write a parser for plain text.

How should we go about compiling a list of suggestions? Obviously it can be expanded over time but some more now might be good. Maybe also a tidy check for this because if you have bad JSON then it'll only get picked up on at runtime (not tested)? And some kind of check to make sure that files matching the globs *do* exist, to keep the suggestions in date?

File: src/bootstrap/suggest.rs
Comment (albertlarsan68) [inline src/bootstrap/suggest.rs:35]: You can get this from [src/bootstrap/format.rs:81-101](https://github.com/rust-lang/rust/blob/master/src/bootstrap/format.rs#L81-L101), and it is a `git` invocation less. Maybe again wait for #105058 to be merged and move it to `src/tools/build-helper/src/git.rs`

File: src/bootstrap/suggest.rs
Comment (albertlarsan68) [reply src/bootstrap/suggest.rs:17]: Did you mean #105058 ?

File: src/bootstrap/suggest.rs
Comment (albertlarsan68) [reply src/bootstrap/suggest.rs:35]: You should also wait for 106xxx

Comment (Ezrashaw) [reply src/bootstrap/suggest.rs:35]: Everything seems quite up in the air. #105058 is getting reverted. I think I'll just wait for this to blow over a bit.

File: src/bootstrap/builder.rs
Comment (Ezrashaw) [reply src/bootstrap/builder.rs:870]: Hmm, using `serde_json` means that we can't directly create `Subcommand`s. Honestly, the solution is good enough as is for now.

**Comment by Ezrashaw** [comment]
@jyn514 I've implemented the things we've talked about. I think we should just wait a bit while the reverts happen. It might turn out that this PR brings in the `build_helper` crate with the git functions. Also see my last comment about the list of suggestions.

**Comment by Ezrashaw** [comment]
@jyn514 I've moved the tool into its own directory and created a "dynamic suggestions" structure with that `compiler/` example. If there's nothing else, then I think it's done?

**Comment by jyn514** [comment]
After thinking about this more, I am not sure how useful this tool will be if new contributors don't know to use it :/ I continued the discussion on https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/pain.20points.20of.20bootstrap/near/319094297 about how to make this more useful.

To be clear, there's nothing wrong with your implementation (I have some minor review comments that aren't blocking), but I think the original idea I had a year ago may not be very useful in practice.

**Comment by Ezrashaw** [comment]
@jyn514 Rebased and squashed. I was having some slight issues with that merge conflict but they should be fixed. After CI passes, should be good to go.

File: src/bootstrap/builder.rs
Comment (jyn514) [reply src/bootstrap/builder.rs:889]: I'm not sure I understand the problem, sorry - how is serde_json related here?

File: src/tools/suggest-tests/src/lib.rs
Comment (jyn514) [inline src/tools/suggest-tests/src/lib.rs:59]: The output here is not great if there are 0 suggestions: maybe we can print something like "No suggestions found" and then either "Consider adding a new rule" or "Modify at least 1 file to get suggestions"?
```
=== SUGGESTIONS ===
===================
help: consider using the `--run` flag to automatically run suggested tests
Build completed successfully in 0:00:13
```

File: src/tools/suggest-tests/src/dynamic_suggestions.rs
Comment (jyn514) [inline src/tools/suggest-tests/src/dynamic_suggestions.rs:30]: This looks like stray debugging, please remove it.
```
[Normal("compiler"), Normal("rustc_borrowck")]
Suggestion { cmd: "test", stage: None, paths: ["compiler/rustc_borrowck"] }
```

File: src/bootstrap/Cargo.toml
Comment (jyn514) [inline src/bootstrap/Cargo.toml:55]: This isn't what I meant - it still rebuilds bootstrap when you modify the code. I was imagining that you would compile it separately and then parse the paths it outputs in bootstrap. Right now your `bootstrap_tool!` line in `tool.rs` is never used.
```
(bash@dev-desktop-us-1.infra.rust-lang.org) ~/rust3 [18:10:09]
> touch src/tools/suggest-tests/src/static_suggestions.rs 
(bash@dev-desktop-us-1.infra.rust-lang.org) ~/rust3 [18:10:16]
> x suggest --run
Building rustbuild
   Compiling suggest-tests v0.1.0 (/home/gh-jyn514/rust3/src/tools/suggest-tests)
   Compiling bootstrap v0.0.0 (/home/gh-jyn514/rust3/src/bootstrap)
    Finished dev [unoptimized] target(s) in 3.90s
```

**Review by jyn514 — COMMENTED** [review]
This looks like a good start! I have just a few nits and then I think we can merge it :) and have people add new suggestions over time.

**Comment by Ezrashaw** [comment]
@jyn514 I've been taking a bit of a break from Rust development (I felt a bit burnt out), I'll try to implement your suggestions early next week. Sorry for the inconvenience.

**Comment by jyn514** [comment]
no worries at all! please take as much time as you need :)

**Comment by Ezrashaw** [comment]
@jyn514 I think that this is ready for another review round. It's kind of drafty but I think I've implemented your suggestions and it seems to work.

**Comment by jyn514** [comment]
i don't have energy for reviews right now, sorry :(

r? bootstrap

File: src/bootstrap/suggest.rs
Comment (albertlarsan68) [inline src/bootstrap/suggest.rs:19]: Can you be more verbose? (printing stderr, stdout and exit code value)

File: src/bootstrap/suggest.rs
Comment (albertlarsan68) [inline src/bootstrap/suggest.rs:38]:
```suggestion
            print!("x {} ", sug.0);
```
So that it prints a "real" command

**Review by albertlarsan68 — CHANGES_REQUESTED** [review]
Also, I suggest you use a sentinel value (e.g. `N/A`) for when there is no stage, to avoid loosing the last path.

**Comment by Ezrashaw** [comment]
@albertlarsan68 Sorry for the slow response. I've implemented all your feedback.

**Review by albertlarsan68 — CHANGES_REQUESTED** [review]
Can you also add tests to the new crate?

**Comment by albertlarsan68** [comment]
If you need guidance on which tests to implement/how to implement them, feel free to ping me in Zulip.

**Comment by Ezrashaw** [comment]
@albertlarsan68 Implemented your feedback. I'm not entirely convinced of the need for tests but I've added them here.

**Review by albertlarsan68 — APPROVED** [review]
This PR is ready to be merged, but there will be some follow-up steps:
- Open an issue to fill the recommendations (and tag T-compiler, T-lib)
- Add documentation for this in the rustc-dev-guide

cc @jyn514 Are you happy with this implementation?
r=me when appropriate.

**Comment by Ezrashaw** [comment]
@jyn514 @albertlarsan68 Sorry, this has just been sitting here for a bit. Am I meant to do something or has jyn just not got to this yet? I'm relatively new to rustc so I don't understand these nuances.

File: src/bootstrap/lib.rs
Comment (jyn514) [inline src/bootstrap/lib.rs:663]: It looks like the `paths` field is never used? Can you remove it from `Subcommand::Suggest`?

Also, this doesn't download rustfmt before running suggest - we should do that here, maybe just by moving the call on line 668 to 657.

File: src/bootstrap/suggest.rs
Comment (jyn514) [inline src/bootstrap/suggest.rs:23]: This should fail the build.
```suggestion
        panic!("failed to run `suggest-tests`");
```

File: src/bootstrap/suggest.rs
Comment (jyn514) [inline src/bootstrap/suggest.rs:34]: I'm having a hard time imagining what the original `line` looked like. Can you leave a comment of what format this code expects?

File: src/bootstrap/suggest.rs
Comment (jyn514) [inline src/bootstrap/suggest.rs:49]:
```suggestion
            for path in &sug.2 {
                print!("{} ", path.display());
```

File: src/bootstrap/suggest.rs
Comment (jyn514) [reply src/bootstrap/suggest.rs:55]: @Ezrashaw looks like this comment was never addressed?

File: src/bootstrap/builder.rs
Comment (jyn514) [reply src/bootstrap/builder.rs:890]: @Ezrashaw I don't think I ever got an answer here - how is serde related?

File: src/tools/suggest-tests/src/lib.rs
Comment (jyn514) [inline src/tools/suggest-tests/src/lib.rs:95]: nit: this is uncommon, it's more traditional to put imports at the top of the file. If the problem is that `sug!` isn't defined at the time you want to re-export it, you can move it up too.

File: src/tools/suggest-tests/src/static_suggestions.rs
Comment (jyn514) [inline src/tools/suggest-tests/src/static_suggestions.rs:25]:
```suggestion
    "src/librustdoc/*" => [
```

File: src/tools/suggest-tests/src/static_suggestions.rs
Comment (jyn514) [inline src/tools/suggest-tests/src/static_suggestions.rs:18]: Can you move this to a dynamic suggestion so it can be specific about which crate to test? Any of the following should work:
```
    ./x.py test library/alloc
    ./x.py test library/core
    ./x.py test library/panic_abort
    ./x.py test library/panic_unwind
    ./x.py test library/proc_macro
    ./x.py test library/std
    ./x.py test library/test
    ./x.py test library/unwind
```

hmm, now I'm wondering again why we chose to make this a separate tool instead of using reflection from the `should_run` functions ... do you happen to have a pointer to that discussion?

File: src/tools/suggest-tests/src/tests.rs
Comment (jyn514) [inline src/tools/suggest-tests/src/tests.rs:21]: nit: this would be easier to read if you wrote it as a for loop inside a single `#[test]` function instead of a macro.

**Review by jyn514 — COMMENTED** [review]
Left lots of little comments, but the only ones I really care about are the question about `builder::new_internal` and why there's a separate `suggest-tests` crate.

File: src/bootstrap/builder.rs
Comment (Ezrashaw) [reply src/bootstrap/builder.rs:890]: > Hmm, using serde_json means that we can't directly create Subcommands. Honestly, the solution is good enough as is for now.

Sorry about that. I have no idea what I was thinking lol. I think that this isn't that relevant anymore since the code was moved to a separate crate.

File: src/bootstrap/suggest.rs
Comment (Ezrashaw) [reply src/bootstrap/suggest.rs:55]: I believe it was? Line 56?

File: src/tools/suggest-tests/src/static_suggestions.rs
Comment (Ezrashaw) [reply src/tools/suggest-tests/src/static_suggestions.rs:18]: I've addressed your first point.

In terms of the second point, I can't seem to find a good place to link to but the general point is that since suggestions are implemented in code (so they can be dynamic and we don't need `serde_json` or whatever), any time a suggestion is updated the code must rebuilt. Having to rebuild bootstrap for a minor suggestion change is too expensive so it was decided that bootstrap would just have a small unchanging shim which would parse a separate crate's output.

File: src/tools/suggest-tests/src/tests.rs
Comment (Ezrashaw) [reply src/tools/suggest-tests/src/tests.rs:21]: Personally I like the idea of having one test per suggestion, it's clearly structured and easier to tell what test failed. Also I think it's quite clear to read, it's just a function with placeholders effectively.

File: src/tools/suggest-tests/src/lib.rs
Comment (Ezrashaw) [reply src/tools/suggest-tests/src/lib.rs:95]: Don't know how those got there lol.

**Comment by Ezrashaw** [comment]
@jyn514 Implemented your feedback and provided answers to your questions.

File: src/tools/suggest-tests/src/tests.rs
Comment (albertlarsan68) [reply src/tools/suggest-tests/src/tests.rs:21]: The macro could generate a single function, with a for loop.

Comment (Ezrashaw) [reply src/tools/suggest-tests/src/tests.rs:21]: But then you don't get the nice "function `test_suggestion_xyz` failed"?

**Comment by Ezrashaw** [comment]
@jyn @albertlarsan68 Rebased, no changes of note.

File: src/tools/suggest-tests/src/static_suggestions.rs
Comment (jyn514) [reply src/tools/suggest-tests/src/static_suggestions.rs]: Hmm, ok. I'm trying to remember why we decided to have "dynamic suggestions" in the first place instead of using reflection in builder.rs. But it doesn't need to block this PR; it's already quite useful and we can improve this in a follow-up.

**Comment by compiler-errors** [comment]
This failed in a rollup: https://github.com/rust-lang-ci/rust/actions/runs/4667818566/jobs/8264150865

```
error[E0277]: the trait bound `BuildMetrics: Clone` is not satisfied
   --> lib.rs:244:5
    |
193 | #[derive(Clone)]
    |          ----- in this derive macro expansion
...
244 |     metrics: metrics::BuildMetrics,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `BuildMetrics`
    |
    = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
```

@bors r-

**Comment by compiler-errors** [comment]
This failed in a rollup: https://github.com/rust-lang-ci/rust/actions/runs/4667818566/jobs/8264150865

```
error[E0277]: the trait bound `BuildMetrics: Clone` is not satisfied
   --> lib.rs:244:5
    |
193 | #[derive(Clone)]
    |          ----- in this derive macro expansion
...
244 |     metrics: metrics::BuildMetrics,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `BuildMetrics`
    |
    = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
```

@bors r-

**Comment by Ezrashaw** [comment]
@compiler-errors Could I get a re-roll please? Issue stemmed from using a CI-only feature.

**Comment by albertlarsan68** [comment]
IIIRC, build metrics are enabled in all CI jobs. I'm not happy with the solution you found, but making this work with build metrics can be done as a follow-up.
@bors r=jyn514,albertlarsan68

**Comment by Ezrashaw** [comment]
@albertlarsan68 

> IIIRC, build metrics are enabled in all CI jobs. I'm not happy with the solution you found, but making this work with build metrics can be done as a follow-up.

Keep in mind that CI should never need to run `x suggest`; it is a user-only feature.

Unfortunately, upstream changes are required for `BuildMetrics` to implement `Clone` in the `sysinfo` crate.

**Pull request merged**
